### PR TITLE
[ARTP-163] Provide localized theming and to-spec theme

### PR DESCRIPTION
**Pull request by evanrs**
I am merging into David's branch to highlight this approach first — before merging all changes together into `story/theming`.

### CODE REVIEW

File: src/client/src/ui/StatusBar/StatusBar.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  resolveTheme = theme => {
+    const { mode } = this.props
+    const currentTheme = theme.statusBar[mode]
+
+    theme = {
+      ...theme,
+      ...theme.statusBar,
+      ...theme.statusBar[mode],
+      serviceItem: {
+        connecting: theme.statusBar.connected,
```
Comment (evanrs) [inline src/client/src/ui/StatusBar/StatusBar.js]: Made a consistent mistake in two places, should be `connected`

File: src/client/src/ui/StatusBar/StatusBar.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            ))}
+          </ServiceList>
+        </Root>
+      </ThemeProvider>
+    )
+  }
+}
+
+const Service = ({ service: { serviceType, isConnected, connectedInstanceCount }, index }) => (
+  <ThemeProvider theme={theme => theme.serviceItem[isConnected ? 'connecting' : 'disconnected']}>
```
Comment (evanrs) [inline src/client/src/ui/StatusBar/StatusBar.js]: Should be `connected`

File: src/client/src/shell/Shell.tsx [hunk 3]
```diff
@@ -98,9 +75,56 @@ class Shell extends React.Component<Props, State> {
...
           )}
         </Transition>
-      </StyledShell>
+      </ShellRoot>
     )
   }
 }
 
+const ShellRoot = styled('div')`
+  width: 100vw;
```
[the remark this answers is not in the document]
Comment (evanrs) [reply src/client/src/shell/Shell.tsx]: Nope, random casualty — will revert.

File: src/client/src/shell/ShellContainer.tsx [hunk 4]
```diff
@@ -10,8 +10,10 @@ import { ConnectionStatus } from 'system'
 import Shell from './Shell'
 
 const mapStateToProps = (state: GlobalState) => ({
+  connected: state.connectionStatus.status === 'connected',
+  loaded: Object.keys(state.compositeStatusService).length >= 3,
```
[the remark this answers is not in the document]
Comment (evanrs) [reply src/client/src/shell/ShellContainer.tsx]: Yes, this would be nicer — the list of services fills up randomly as if the app has no idea what to expect.

File: src/client/src/ui/StatusBar/StatusBar.stories.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    </Story>
+  </Provider>
+)
+
+const Container = styled.div`
+  display: flex;
+  align-items: flex-end;
+  height: 100%;
+  width: 100%;
+`
```
[the remark this answers is not in the document]
Comment (evanrs) [reply src/client/src/ui/StatusBar/StatusBar.stories.js]: The flex component isn't very useful — especially if I'm creating a styled component. It may be useful for a quick sketch, but not in combination.

Since we won't be reinterpreting the flex values into some special subset of properties the `Flex` component adds an uninteresting layer of indirection.

File: src/client/src/ui/blotter/reducer.ts [hunk 6]
```diff
@@ -12,22 +13,22 @@ export interface BlotterState {
...
-      const newTradesById = keyBy(action.payload.trades, `tradeId`)
+      const newTradesById = _.keyBy(action.payload.trades, `tradeId`)
       return {
         ...state,
         trades: {
           ...state.trades,
-          ...newTradesById
-        }
+          ...newTradesById,
+        },
```
[the remark this answers is not in the document]
Comment (evanrs) [reply src/client/src/ui/blotter/reducer.ts]: `trailingComma: "all"`!! I say `"all"`! 😝

File: src/client/.storybook/webpack.config.js [hunk 7]
```diff
@@ -1,22 +1,26 @@
...
     'rt-components': path.resolve(__dirname, '../src', 'rt-components'),
     'rt-util': path.resolve(__dirname, '../src', 'rt-util'),
     'rt-types': path.resolve(__dirname, '../src', 'rt-types'),
     'rt-actions': path.resolve(__dirname, '../src', 'rt-actions'),
     'rt-storybook': path.resolve(__dirname, '../src', 'rt-storybook'),
-    'rt-style-guide': path.resolve(__dirname, '../src', 'rt-style-guide'),
+    'rt-styleguide': path.resolve(__dirname, '../src', 'rt-styleguide'),
     'rt-themes': path.resolve(__dirname, '../src', 'rt-themes'),
+    'rt-theme': path.resolve(__dirname, '../src', 'rt-theme'),
     ui: path.resolve(__dirname, '../src', 'ui'),
```
Comment (bhavesh-desai-scratch) [inline src/client/.storybook/webpack.config.js]: I do not think we should reference UI, Shell and system using aliases. I will put a task to remove these in a seperate PR

File: src/client/config-overrides.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+const { compose, getBabelLoader, injectBabelPlugin } = require('react-app-rewired')
+
+module.exports = function override(config, env) {
```
Comment (bhavesh-desai-scratch) [inline src/client/config-overrides.js]: We comment this file to state what it does.

File: src/client/src/rt-theme/colors.js [hunk 9]
```diff
@@ -0,0 +1,203 @@
+import _ from 'lodash';
```
Comment (bhavesh-desai-scratch) [inline src/client/src/rt-theme/colors.js]: No more loadash!, Also if you really need it just pull what you need

File: src/client/package.json [hunk 10]
```diff
@@ -44,7 +44,7 @@
   },
   "license": "MIT",
   "lint-staged": {
-    "src/**/*.{ts,tsx,json,scss}": [
+    "src/**/*.{js,ts,tsx,json,scss}": [
```
Comment (bhavesh-desai-scratch) [inline src/client/package.json]: We do not want to do this. No JS on this project, please.

File: src/client/src/rt-theme/colors.js (hunk 9 above)
Comment (bhavesh-desai-scratch) [inline src/client/src/rt-theme/colors.js]: Can you make this a TS file, please. Like the rest of the application

File: src/client/src/rt-theme/colors.js [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    }, {}),
+
+    // darks
+    ..._.range(10, 0).reduce((acc, degree) => {
+      acc[`D${degree}`] = mix(degree / 10, black, base);
+
+      return acc;
+    }, {}),
+
+    // transparents
```
Comment (bhavesh-desai-scratch) [inline src/client/src/rt-theme/colors.js]: Why check in commented code?

File: src/client/src/rt-theme/tools.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import _ from 'lodash'
```
Comment (bhavesh-desai-scratch) [inline src/client/src/rt-theme/tools.ts]: Hard to follow! Also things like this should be tested

File: src/client/src/rt-theme/colors.js [hunk 13]
```diff
@@ -0,0 +1,203 @@
+import _ from 'lodash';
+import { mix, rgb, rgba } from 'polished';
```
Comment (bhavesh-desai-scratch) [inline src/client/src/rt-theme/colors.js]: Can you please provide some comments on this file to describe your approach. This is a very important part of the application and developers will need to understand the core concepts

File: src/client/src/rt-theme/tools.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+export const resolvesColor = (color, other) => {
```
Comment (bhavesh-desai-scratch) [inline src/client/src/rt-theme/tools.ts]: Are you even using this?

File: src/client/src/rt-theme/index.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+export { colors } from './colors'
```
Comment (bhavesh-desai-scratch) [inline src/client/src/rt-theme/index.ts]: Do we still need RT-themes and rt-theme?

File: src/client/src/rt-theme/tools.ts [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+    if (value[0] === '#' || _.startsWith(value, 'rgb') || _.startsWith(value, 'hsl')) {
+      return value
+    }
+
+    value = resolvesColor(value, null)(props) || (other ? resolvesColor(other, null)(props) : null)
+
+    return value
+  }
+}
```
[the remark this answers is not in the document]
Comment (evanrs) [reply src/client/src/rt-theme/tools.ts]: Yes, unfortunately I'd added flow types — those help. I'll be adding typescript types and we'll see if that helps clarify the behavior.

File: src/client/src/rt-theme/index.ts (hunk 15 above)
Comment (evanrs) [reply src/client/src/rt-theme/index.ts]: We don't, this is transitory — want to get this merged for @keshaldesai. My next PR addresses this.

File: src/client/src/rt-theme/tools.ts (hunk 14 above)
Comment (evanrs) [reply src/client/src/rt-theme/tools.ts]: No, not at the moment, casualty of copy paste — will remove.

File: src/client/src/rt-theme/colors.js (hunk 13 above)
Comment (evanrs) [reply src/client/src/rt-theme/colors.js]: Yes I will document the what and why with regards to colors and palettes.

Are you also suggesting I document some color theory or how the math from these libs is incorrect?

File: src/client/src/rt-theme/tools.ts (hunk 12 above)
Comment (evanrs) [reply src/client/src/rt-theme/tools.ts]: Yes, but their behavior is brilliant — this is an oldie but a goodie, wrote it two years ago. But still, not using, so I'm removing it for now.

File: src/client/src/rt-theme/colors.js (hunk 11 above)
Comment (evanrs) [reply src/client/src/rt-theme/colors.js]: Casualty of copied code, good catch.

File: src/client/src/rt-theme/colors.js (hunk 9 above)
Comment (evanrs) [reply src/client/src/rt-theme/colors.js]: Let's pair on this.

File: src/client/package.json (hunk 10 above)
Comment (evanrs) [reply src/client/package.json]: Because all code should be formatted — no matter what it is. Formatting code isn't an endorsement it's a comprehensive policy.

File: src/client/src/rt-theme/colors.js (hunk 9 above)
Comment (evanrs) [reply src/client/src/rt-theme/colors.js]: That's build step `babel-plugin-lodash`. In answer, no.

File: src/client/config-overrides.js (hunk 8 above)
Comment (evanrs) [reply src/client/config-overrides.js]: I can add a link to https://github.com/timarney/react-app-rewired — where proper documentation can be found.

If you don't know what a webpack config is or how babel works then the reader has homework. The level of explanation depends on the goal with the reader. A typical code base has a high expectation of the readers contextual understanding.

This is not to mention how quickly this writing grows stale — and as it is outside the critical path of the application it typically lacks an owner to carry it forward.

File: src/client/.storybook/webpack.config.js (hunk 7 above)
Comment (evanrs) [reply src/client/.storybook/webpack.config.js]: Sounds good to me.

File: src/client/src/rt-theme/ThemeProvider.js [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+      isEqual(theme, state.theme)
+      ? null
+      : {
+          theme: {
+            // TODO explore how to protect values
+            ...parentTheme,
+            ...theme
+          }
+        }
+  }
```
Comment (keshaldesai) [inline src/client/src/rt-theme/ThemeProvider.js]: [https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization)

Comment (evanrs) [reply src/client/src/rt-theme/ThemeProvider.js]: But do you see why we do need `getDerivedStateFromProps` here?

Try writing an alternate approach and we can go over it.
